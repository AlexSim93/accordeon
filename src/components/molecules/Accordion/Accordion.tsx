import React, { useState, useCallback } from 'react';
import noop from 'lodash/noop';
import { Chevron } from '../../../assets/icons';
import { ICON_SIZE, Icon } from '../../atoms/Icon';
import { Divider } from '../../atoms/Divider';
import { Text } from '../../atoms/Text';
import { AccordionContainer } from './components/AccordionContainer';
import { TitleContainer } from './components/TitleContainer';
import { DescriptionContainer } from './components/DecriptionContainer';
import { AnimatedChevron } from './components/AnimatedChevron';

type Props = {
  title: string;
  expanded?: boolean;
  initiallyExpanded?: boolean;
  onToggle?: (state: boolean) => void;
};

export const Accordion: React.FC<Props> = ({
  title,
  initiallyExpanded = false,
  expanded,
  onToggle = noop,
  children,
}) => {
  const [isExpanded, toggleAccordion] = useState<boolean>(initiallyExpanded);

  const handleToggleAccordion = useCallback(() => {
    toggleAccordion(!isExpanded);
    onToggle(isExpanded);
  }, [isExpanded, onToggle]);

  const shouldBeExpanded = typeof expanded === 'boolean' ? expanded : isExpanded;

  return (
    <AccordionContainer
      bg="darkGreyLighten16"
      fontFamily="Montserrat"
      borderRadius="s"
    >
      <TitleContainer
        p="xs"
        flexDirection="row"
        alignItems="center"
        onClick={handleToggleAccordion}
      >
        <Text color="white" fontSize={['s', 'm', 'm']} flexGrow={1}>
          {title}
        </Text>
        <AnimatedChevron shouldBeExpanded={shouldBeExpanded}>
          <Icon type={Chevron} size={ICON_SIZE.s} fill="#C4C4C4" />
        </AnimatedChevron>
      </TitleContainer>
      {shouldBeExpanded && <Divider height={1} bg="blackPearlLighten48" />}
      {shouldBeExpanded && (
        <DescriptionContainer
          color="medianGrey"
          px="xs"
          pt="xs"
          pb={['xs', 's', 's']}
        >
          {children}
        </DescriptionContainer>
      )}
    </AccordionContainer>
  );
};
