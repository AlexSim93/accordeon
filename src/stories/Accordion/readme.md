# Accordion

Используется для отображение дополнительной информации на клик по заголовку. На повторный клик происходит закрытие.

## Свойства:

#### Accordion

| Свойство          | Тип TS                   | required | default | Описание                                             |
| ----------------- | ------------------------ | -------- | ------- | ---------------------------------------------------- |
| initiallyExpanded | boolean                  | ✗        | false   | состояние аккордеона, по умолчанию false             |
| title             | string                   | ✓        | -       | заголовок аккордеона                                 |
| onToggle          | (state: boolean) => void | ✗        | noop    | коллбэк, который будет вызван при клике на заголовке |
| expanded          | boolean                  | ✗        | -       | Пропса для управления аккордеоном извне              |

## Использование:

- Аккордеон { title="title", initiallyExpanded=true }

```js
<Accordion title="title" initiallyExpanded>
  Description
</Accordion>
```
