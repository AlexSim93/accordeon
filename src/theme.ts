const breakpoints = ['768px', '1440px'];

export default {
  breakpoints,
  fontSizes: {
    xxs: '10px',
    xs: '12px',
    s: '14px',
    m: '16px',
    ml: '20px',
    l: '24px',
    xl: '32px',
    xxl: '40px',
    xxxl: '56px',
  },
  space: { // s 8, m 16, l 24
    xxxxs: '4px',
    xxxs: '8px',
    xxs: '12px',
    xs: '16px',
    s: '24px',
    sm: '32px',
    m: '40px',
    ml: '48px',
    l: '56px',
    xl: '64px',
    xxl: '72px',
    xxxl: '140px',
    xxxxl: '200px',
  },
  mediaQueries: {
    small: `@media screen and (min-width: ${breakpoints[0]})`,
    medium: `@media screen and (min-width: ${breakpoints[1]})`,
  },
  colors: {
    white: '#FFFFFF', // title text
    blackPearlLighten48: 'rgba(8, 13, 38, 0.48)', // border
    medianGrey: '#A7A7C8', // description text
    darkGreyLighten16: 'rgba(103, 107, 139, 0.16)', // accordion bg
    neutralDarkGrey: '#C4C4C4', // icon
    darkBlue: '#080D26', // page bg
    dodgerBlue: '#564FFF', // button bg
  },
  fonts: {
    Montserrat: '"Montserrat", sans-serif',
  },
  radii: {
    s: '4px',
  },
};
