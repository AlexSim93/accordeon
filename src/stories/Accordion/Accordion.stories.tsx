import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { Box, Flex } from 'reflexbox';
import { ThemeProvider } from 'styled-components';
import { Text } from '../../components/atoms/Text';
import theme from '../../theme';
import { Accordion } from '../../components/molecules/Accordion';
// @ts-ignore
import notes from './readme.md';
import { Paragraph } from '../../components/atoms/Paragraph';

storiesOf('Accordion', module).add(
  'Accordion',
  () => {
    const [isExpanded, changeState] = useState(false);
    return (
      <ThemeProvider theme={theme}>
        <Flex flexDirection="column">
          <Text color="black" fontWeight="600" fontSize="xl">
            Accordion
          </Text>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion
            </Text>
            <Accordion title="simple accordion">
              <Paragraph>
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum
                fugiat possimus commodi veritatis recusandae esse magni, odit
                iusto ipsa quaerat voluptate deserunt aspernatur officia amet
                repudiandae, perferendis similique repellendus minus.
              </Paragraph>
            </Accordion>
          </Box>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion initially expanded
            </Text>
            <Accordion title="simple accordion" initiallyExpanded>
              <Paragraph>
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum
                fugiat possimus commodi veritatis recusandae esse magni, odit
                iusto ipsa quaerat voluptate deserunt aspernatur officia amet
                repudiandae, perferendis similique repellendus minus.
              </Paragraph>
            </Accordion>
          </Box>
          <Box mb={24}>
            <Text color="black" fontSize="l">
              Accordion with external control
            </Text>
            <Box py={16}>
              <button type="button" onClick={() => changeState(true)}>Expand</button>
              <button type="button" onClick={() => changeState(false)}>Collapse</button>
            </Box>
            <Accordion title="simple accordion" expanded={isExpanded}>
              <Paragraph>
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum
                fugiat possimus commodi veritatis recusandae esse magni, odit
                iusto ipsa quaerat voluptate deserunt aspernatur officia amet
                repudiandae, perferendis similique repellendus minus.
              </Paragraph>
            </Accordion>
          </Box>
        </Flex>
      </ThemeProvider>
    );
  },
  { notes },
);
