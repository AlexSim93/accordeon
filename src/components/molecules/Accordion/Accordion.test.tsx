import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { Accordion } from './Accordion';
import theme from '../../../theme';

describe('Accordion', () => {
  it('Accordion should render correctly by default', () => {
    const accordion = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Accordion title="title">Description</Accordion>
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordion).toMatchSnapshot();
  });

  it('Accordion should render correctly when initiallyExpanded=true', () => {
    const accordion = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Accordion title="title" initiallyExpanded>
            Description
          </Accordion>
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordion).toMatchSnapshot();
  });

  it('Accordion should render correctly when expanded=true', () => {
    const accordion = renderer
      .create(
        <ThemeProvider theme={theme}>
          <Accordion title="title" expanded>
            Description
          </Accordion>
        </ThemeProvider>,
      )
      .toJSON();
    expect(accordion).toMatchSnapshot();
  });

  it('Accordion should expand and collapse on click', () => {
    const accordion = mount(<Accordion title="title">Description</Accordion>);

    const accordionTitle = accordion.find('TitleContainer');

    accordionTitle.simulate('click');

    expect(accordion.exists('DescriptionContainer')).toEqual(true);

    accordionTitle.simulate('click');

    expect(accordion.exists('DescriptionContainer')).toEqual(false);
  });
});
