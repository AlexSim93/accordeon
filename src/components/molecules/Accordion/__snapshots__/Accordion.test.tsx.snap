// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Accordion should render correctly by default 1`] = `
.c2 {
  font-size: 14px;
  font-family: "Montserrat",sans-serif;
  color: #FFFFFF;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  background-color: rgba(103,107,139,0.16);
  font-family: "Montserrat",sans-serif;
  border-radius: 4px;
  font-style: normal;
}

.c1 {
  padding: 16px;
  font-weight: 600;
}

.c1:hover {
  cursor: pointer;
}

.c3 {
  -webkit-transform: rotate( 360deg );
  -ms-transform: rotate( 360deg );
  transform: rotate( 360deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

@media screen and (min-width:768px) {
  .c2 {
    font-size: 16px;
  }
}

@media screen and (min-width:1024px) {
  .c2 {
    font-size: 16px;
  }
}

<div
  className="c0 css-nwbphr"
>
  <div
    className="c1 css-vpgf7p"
    onClick={[Function]}
  >
    <span
      className="c2"
      color="white"
      fontFamily="Montserrat"
      fontSize={
        Array [
          "s",
          "m",
          "m",
        ]
      }
    >
      title
    </span>
    <div
      className="c3 css-vurnku"
    />
  </div>
</div>
`;

exports[`Accordion Accordion should render correctly when expanded=true 1`] = `
.c4 {
  height: 1px;
  background-color: rgba(8,13,38,0.48);
}

.c2 {
  font-size: 14px;
  font-family: "Montserrat",sans-serif;
  color: #FFFFFF;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  background-color: rgba(103,107,139,0.16);
  font-family: "Montserrat",sans-serif;
  border-radius: 4px;
  font-style: normal;
}

.c1 {
  padding: 16px;
  font-weight: 600;
}

.c1:hover {
  cursor: pointer;
}

.c5 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 16px;
  padding-bottom: 16px;
  color: #A7A7C8;
  font-weight: 500;
}

.c3 {
  -webkit-transform: rotate( 180deg );
  -ms-transform: rotate( 180deg );
  transform: rotate( 180deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

@media screen and (min-width:768px) {
  .c2 {
    font-size: 16px;
  }
}

@media screen and (min-width:1024px) {
  .c2 {
    font-size: 16px;
  }
}

@media screen and (min-width:768px) {
  .c5 {
    padding-bottom: 24px;
  }
}

@media screen and (min-width:1024px) {
  .c5 {
    padding-bottom: 24px;
  }
}

<div
  className="c0 css-nwbphr"
>
  <div
    className="c1 css-vpgf7p"
    onClick={[Function]}
  >
    <span
      className="c2"
      color="white"
      fontFamily="Montserrat"
      fontSize={
        Array [
          "s",
          "m",
          "m",
        ]
      }
    >
      title
    </span>
    <div
      className="c3 css-vurnku"
    />
  </div>
  <div
    className="c4"
    height={1}
  />
  <div
    className="c5 css-th0e4m"
  >
    Description
  </div>
</div>
`;

exports[`Accordion Accordion should render correctly when initiallyExpanded=true 1`] = `
.c4 {
  height: 1px;
  background-color: rgba(8,13,38,0.48);
}

.c2 {
  font-size: 14px;
  font-family: "Montserrat",sans-serif;
  color: #FFFFFF;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  background-color: rgba(103,107,139,0.16);
  font-family: "Montserrat",sans-serif;
  border-radius: 4px;
  font-style: normal;
}

.c1 {
  padding: 16px;
  font-weight: 600;
}

.c1:hover {
  cursor: pointer;
}

.c5 {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 16px;
  padding-bottom: 16px;
  color: #A7A7C8;
  font-weight: 500;
}

.c3 {
  -webkit-transform: rotate( 180deg );
  -ms-transform: rotate( 180deg );
  transform: rotate( 180deg );
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

@media screen and (min-width:768px) {
  .c2 {
    font-size: 16px;
  }
}

@media screen and (min-width:1024px) {
  .c2 {
    font-size: 16px;
  }
}

@media screen and (min-width:768px) {
  .c5 {
    padding-bottom: 24px;
  }
}

@media screen and (min-width:1024px) {
  .c5 {
    padding-bottom: 24px;
  }
}

<div
  className="c0 css-nwbphr"
>
  <div
    className="c1 css-vpgf7p"
    onClick={[Function]}
  >
    <span
      className="c2"
      color="white"
      fontFamily="Montserrat"
      fontSize={
        Array [
          "s",
          "m",
          "m",
        ]
      }
    >
      title
    </span>
    <div
      className="c3 css-vurnku"
    />
  </div>
  <div
    className="c4"
    height={1}
  />
  <div
    className="c5 css-th0e4m"
  >
    Description
  </div>
</div>
`;
